src/EditorFeatures/Core/Implementation/NavigationBar/NavigationBarController_ModelComputation.cs

```diff
@@ -142,11 +141,15 @@ private void StartSelectedItemUpdateTask(int delay, bool updateUIWhenDone)
             if (updateUIWhenDone)
             {
                 asyncToken = _asyncListener.BeginAsyncOperation(GetType().Name + ".StartSelectedItemUpdateTask.UpdateUI");
-                _selectedItemInfoTask.SafeContinueWith(
-                    t => PushSelectedItemsToPresenter(t.Result),
+                _selectedItemInfoTask.SafeContinueWithFromAsync(
+                    async t =>
+                    {
+                        await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(alwaysYield: true, cancellationToken);
+                        PushSelectedItemsToPresenter(t.Result);
+                    },
                     cancellationToken,
-                    TaskContinuationOptions.OnlyOnRanToCompletion,
-                    ForegroundTaskScheduler).CompletesAsyncOperation(asyncToken);
+                    TaskContinuationOptions.OnlyOnRanToCompletion | TaskContinuationOptions.ExecuteSynchronously,
+                    TaskScheduler.Default).CompletesAsyncOperation(asyncToken);
```

Comment:
this entire pattern seems repeated several times.  can we extract out a helper extension that does all of this for us.  That way we can make sure no one forgets an option, or improperly passes the wrong value to alwaysYield and whatnot?


effectively, it looks like it would be:

```c#
public static ContinueOnMainThreadAsync<T>(this Task<T> task, Action<T> act) { ... }
```

Reply:
➡️ This pattern is the subject of the ongoing discussion in https://github.com/dotnet/roslyn/pull/28781#discussion_r208378244.

> can we extract out a helper extension

Wrappers around core vs-threading functionality (in this case `SwitchToMainThreadAsync`) are discouraged. The *long* term solution will be to migrate the code to the standard pattern where no helper is needed. The short term solution, if any, is part of the linked discussion.

Reply:
➡️ **tl;dr:** Before adding a wrapper, we should determine if the situation should be addressed by vs-threading itself so the benefit of consistency applies across the range of projects and applications using this library.

A primary goal of vs-threading is to establish a set of primitives, rules/practices, and analyzers, such that interactive applications using asynchronous code can be developed in a reliable, performant, and maintainable manner by developers of mixed skill levels. Achieving this goal across applications involves the creation of patterns that are simple enough to not need an abstraction, and also specific enough that analyzer enforcement with low false-positive and false-negative rates are practical. We should generally be at that point already with the library, which led to a general recommendation that code not be abstracted. In some cases, abstractions will not interfere with the analyzers, but in other cases they do.

We have code which is effectively wrapping or bypassing vs-threading today, but the short term goal would be to discontinue the ongoing use of these patterns (in favor of now-available vs-threading patterns) and phase out their use over time. I've already created some targeted improvements to the vs-threading analyzers to aid in this effort.